API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs

```diff
@@ -0,0 +1,293 @@
+    [Experimental("EXTEXP0012", UrlFormat = "https://aka.ms/dotnet-extensions-warnings/{0}")]
+    public static IServiceCollection AddActivatedKeyedSingleton(this IServiceCollection services, Type serviceType, object? serviceKey, Func<IServiceProvider, object?, object> implementationFactory);
+
+    /// <summary>
+    /// Adds an auto-activated keyed singleton service.
+    /// </summary>
+    /// <param name="services">The service collection to add the service to.</param>
+    /// <param name="serviceType">The type of the service to register.</param>
+    /// <param name="serviceKey">An object used to uniquely identity the specific service.</param>
+    /// <param name="implementationType">The implementation type of the service.</param>
+    /// <returns>The value of <paramref name="services" />.</returns>
+    [Experimental("EXTEXP0012", UrlFormat = "https://aka.ms/dotnet-extensions-warnings/{0}")]
+    public static IServiceCollection AddActivatedKeyedSingleton(this IServiceCollection services, Type serviceType, object? serviceKey, Type implementationType);
+
+    /// <summary>
+    /// Tries to add an auto-activated keyed singleton service.
+    /// </summary>
+    /// <param name="services">The service collection to add the service to.</param>
+    /// <param name="serviceType">The type of the service to register.</param>
+    /// <param name="serviceKey">An object used to uniquely identity the specific service.</param>
+    [Experimental("EXTEXP0012", UrlFormat = "https://aka.ms/dotnet-extensions-warnings/{0}")]
+    public static void TryAddActivatedKeyedSingleton(this IServiceCollection services, Type serviceType, object? serviceKey);
+
+    /// <summary>
+    /// Tries to add an auto-activated keyed singleton service.
+    /// </summary>
+    /// <param name="services">The service collection to add the service to.</param>
+    /// <param name="serviceType">The type of the service to register.</param>
+    /// <param name="serviceKey">An object used to uniquely identity the specific service.</param>
+    /// <param name="implementationType">The implementation type of the service.</param>
+    [Experimental("EXTEXP0012", UrlFormat = "https://aka.ms/dotnet-extensions-warnings/{0}")]
+    public static void TryAddActivatedKeyedSingleton(this IServiceCollection services, Type serviceType, object? serviceKey, Type implementationType);
+
+    /// <summary>
+    /// Tries to add an auto-activated keyed singleton service.
+    /// </summary>
+    /// <param name="services">The service collection to add the service to.</param>
+    /// <param name="serviceType">The type of the service to register.</param>
+    /// <param name="serviceKey">An object used to uniquely identity the specific service.</param>
+    /// <param name="implementationFactory">The factory that creates the service.</param>
+    [Experimental("EXTEXP0012", UrlFormat = "https://aka.ms/dotnet-extensions-warnings/{0}")]
```

Comment:
@jeffhandley is there a central doc where we have to claim/register ID's for [Experimental ] attributes? I thought we had one for BCLnnnn but I can't find that right now.

Reply:
@danmoseley I believe you're thinking of this doc that captures the list of diagnostics (including obsoletions). The page also includes instructions for how we keep the constants organized in the dotnet/runtime codebase.

https://github.com/dotnet/runtime/blob/main/docs/project/list-of-diagnostics.md

We don't have a registration across repos, but dotnet/extensions could introduce the same approach that document promotes. You can also use CODEOWNERS to tag specific team members when new entries are added if desired.

Reply:
We have a simple version of that doc too at https://github.com/dotnet/extensions/blob/main/docs/list-of-diagnostics.md.

Reply:
A while ago I tried to see if there was one central registry for obsoletions and other diagnostics, and I was told that every repo/product maintains their own. 
To list a few:
* https://github.com/dotnet/runtime/blob/main/docs/project/list-of-diagnostics.md?rgh-link-date=2023-09-12T00%3A25%3A27Z
* https://github.com/dotnet/aspnetcore/blob/main/docs/list-of-diagnostics.md
* https://github.com/dotnet/winforms/blob/main/docs/list-of-diagnostics.md